ext/strings_test.go

```diff
@@ -329,3 +331,338 @@ func TestStringsWithExtension(t *testing.T) {
+		},
+		{
+			name:           "lowercase scientific notation formatting clause",
+			format:         "%.6e",
+			formatArgs:     "1052.032911275",
+			expectedOutput: "1.052033\u202f\u00d7\u202f10\u2070\u00b3",
+			locale:         "en_US",
+		},
+		{
+			name:           "uppercase scientific notation formatting clause",
+			format:         "%.6E",
+			formatArgs:     "1052.032911275",
+			expectedOutput: "1.052033\u202f\u00d7\u202f10\u2070\u00b3",
+			locale:         "en_US",
+		},
+		{
+			name:           "locale support",
+			format:         "%.3f",
+			formatArgs:     "3.14",
+			locale:         "fr_FR",
+			expectedOutput: "3,140",
+		},
+		{
+			name:           "default precision for fixed-point clause",
+			format:         "%f",
+			formatArgs:     "2.71828",
+			expectedOutput: "2.718280",
+			locale:         "en_US",
+		},
+		{
+			name:           "default precision for scientific notation",
+			format:         "%e",
+			formatArgs:     "2.71828",
+			expectedOutput: "2.718280\u202f\u00d7\u202f10\u2070\u2070",
+			locale:         "en_US",
+		},
+		{
+			name:           "unicode output for scientific notation",
+			format:         "unescaped unicode: %e, escaped unicode: %e",
+			formatArgs:     "2.71828, 2.71828",
+			expectedOutput: "unescaped unicode: 2.718280 × 10⁰⁰, escaped unicode: 2.718280\u202f\u00d7\u202f10\u2070\u2070",
+			locale:         "en_US",
+		},
+		{
+			name:           "NaN support for fixed-point",
+			format:         "%f",
+			formatArgs:     `"NaN"`,
+			expectedOutput: "NaN",
+			locale:         "en_US",
+		},
+		{
+			name:           "positive infinity support for fixed-point",
+			format:         "%f",
+			formatArgs:     `"Infinity"`,
+			expectedOutput: "∞",
+			locale:         "en_US",
+		},
+		{
+			name:           "negative infinity support for fixed-point",
+			format:         "%f",
+			formatArgs:     `"-Infinity"`,
+			expectedOutput: "-∞",
+			locale:         "en_US",
+		},
+		{
+			name:           "uint support for decimal clause",
+			format:         "%d",
+			formatArgs:     "uint(64)",
+			expectedOutput: "64",
+		},
+		{
+			name:           "timestamp support for string",
+			format:         "%s",
+			formatArgs:     `timestamp("2023-02-03T23:31:20+00:00")`,
+			expectedOutput: "2023-02-03T23:31:20Z",
+		},
+		{
+			name:           "duration support for string",
+			format:         "%s",
+			formatArgs:     `duration("1h45m47s")`,
+			expectedOutput: "6347s",
+		},
+		{
+			name:           "list support for string",
+			format:         "%s",
+			formatArgs:     `["abc", 3.14, [9, 8, 7, 6]]`,
+			expectedOutput: `["abc", "3.140000", ["9", "8", "7", "6"]]`,
+			locale:         "en_US",
+		},
+		{
+			name:           "map support for string",
+			format:         "%s",
+			formatArgs:     `{"key1": b"xyz", "key2": duration("2h"), "key4": true, "key3": 2.71828}`,
+			locale:         "nl_NL",
+			expectedOutput: `{"key1":"eHl6", "key2":"7200s", "key3":"2,718280", "key4":true}`,
```

Comment:
Apologies if I missed this being discussed elsewhere, but what is the rationale for quoting and localizing numbers (rather than representing them as a json number) in map/list output while representing booleans with their proper json type?

Reply:
> Curious if we considered dropping quotes from numbers, since this would be less surprising to the vast majority of JSON users.

I thought about it, though at least for floats/doubles `"Nan"` and `"Infinity"` means they'll be strings at least sometime, so I thought it would be less surprising if they were consistently strings. Integers don't have that issue, so that doesn't necessarily apply to them, though I thought it would be weird if one set of numbers were always strings and another set weren't.

RE: locale/surprise, I guess it depends on the intended use case. (Can protobuf deal with comma decimal separators?) My own opinion is that the option of least surprise would be that the locale applies to all numbers (since the user has to explicitly pass it as an option to the env), but if others disagree, I don't think I'd push back too much.

Reply:
Is the goal to conform to https://protobuf.dev/programming-guides/proto3/#json? This seems reasonable since other languages might then be able to implement this by leveraging existing proto3->json conversion utilities.   If so, I don't think localizing numbers is an option.

For numbers, I'm not convinced. When I see a data dump that is obviously JSON, it is super surprising for me to see types mucked with. When I first saw numbers quoted I was very confused, then I found https://protobuf.dev/programming-guides/proto3/#json and realized it was allowed, but I still don't prefer it. I realize this is part of a a formatting library so there is a temptation to try to do special formatting things, but this is clearly the "data dump" part of that library, and if the data dump format is JSON. Given that I would prefer it be as compliant with JSON as possible. Ideally, I could copy paste the JSON into other tools and it would work as expected. That is not the case once we type swap numbers to strings.

"NaN" and "Infinity" are corner cases that we have to cope with, but for the vast majority of use cases, they are not encountered, so I'm uncomfortable saying that we should make the normal behavior less convenient to accommodate the case. I would just handle those two cases as strings and leave all the other cases as number.

Reply:
Oh, I like the idea of CEL literals.  That would be my preference.

I don't mind if JSON in CEL is formatted as JSON (k8s wouldn't see it, I don't think), I don't have a good sense on if that creates any edge cases.

Reply:
That makes sense to me as well. I'll go ahead and adjust the output such that it's valid CEL.

Reply:
(marking as resolved now that `%s` prints out maps/lists as CEL literals and test cases are included that check that the output is valid CEL)